File: spotify.py
import requests
import time
from pprint import pprint

user_id = 'phig1hflpa6zsgnf36cqhly0p'

ACCESS_TOKEN = 'BQDRax2Pzbj95XrA1N2b25qCbg90wvDXcnUaAK5f8SV_sKl1HVXTwx5j1LRAtVGrDsnnKh6ylo5bRATD6T5M69SLdYVG474mQzBTu5eElDDbVQ39aTAV9R8SyV0iJfEviDTyW7ul_Xrb_3z-RKw6OVg3ksecTh0HZ1jtmJHw86_EWqWfIQT7DqlriUAWtJcvRP2-m_CwB9K2IkDR4DLcL6Z8hbfFRwPF3KZiq7n1eZw-ziqggXFknra2G1PlAhNRRAyxoxf9cwqOWOPqVbG76tJp3avTrn3ZKrGWXAP9'

API_CREATE_PLAYLIST = f'https://api.spotify.com/v1/users/{user_id}/playlists'
API_GET_CURRENT_TRACK = 'https://api.spotify.com/v1/me/player'


# SPOTIFY API FUNCTIONS
def create_playlist(name, public):
    response = requests.post(
        API_CREATE_PLAYLIST,
        headers={
            "Authorization": f"Bearer {ACCESS_TOKEN}"
        },
        json={
            "name": name,
            "public":public
        }
    )

    json_resp = response.json()

    return json_resp

def get_current_track_info():
    response = requests.get(
        API_GET_CURRENT_TRACK,
        headers={
            "Authorization": f"Bearer {ACCESS_TOKEN}"
        }
    )

    # First retrieval error check
    if (response.status_code != 200):
        return response
    else:
        # Convert the response to json
        json_resp = response.json()

    # Second retrieval error check
    if (json_resp.get('error') == None):
        # retrieve the info from the json response
        track_id = json_resp["item"]["id"]
        track_name = json_resp["item"]["name"]
        song_link = json_resp["item"]["external_urls"]["spotify"]
        artist_info = json_resp["item"]["artists"]
        artist_names = ', '.join( [artist["name"] for artist in artist_info] )

        # convert it into a dictionary and return
        current_track_info = {
            "song_id":track_id,
            "song_name":track_name,
            "song_link":song_link,
            "artist_names":artist_names,
        }

        return current_track_info
    else:
        print("NOTE: Probably your Spotify access token is expired...")
        return json_resp.get('error')



# RUN API FUNCTIONS IN MAIN ONLY FOR TESTING PURPOSES!
def main():
    # # Create a playlist
    # playlist = create_playlist(
    #     name = "Test Playlist",
    #     public = False
    # )

    # Get the current track info
    current_track = get_current_track_info()

    # pprint(f'Playlist: {playlist}')
    print()
    pprint(current_track, indent=4)



if __name__ == "__main__":
    main()

File: Spotify/spotify.js
// TODO: These 3 variables are constant for now, but they will dynamicly change according to user 
const user_id = 'phig1hflpa6zsgnf36cqhly0p';
const client_id = 'd61fb49a895346d999907535f080c9c6';
const client_secret = 'e98f203ad55d4ae9884cdecd2ad2b068';

// ! YOU NEED TO UPDATE THIS TOKEN EVERY HOUR or API WILL NOT WORK! (will fix this)
const ACCESS_TOKEN = 'BQADrv0pZCan3Qh705pX94-Wv4cgyr7GL6uo3Gl5bxuKpgtCznjuoLj-iJ5zPwZgwRukORiNE9umSgIus13qmsfQLz-qkjoeOfjJIFyU2U4dVNBDRIuXR24JFBqRbI2ZXQoGYKJzhs6zHO6E0vhS4S-Z22u6JK9k91cJ1xXhV-SzqS1dJBSCxLiaXy-FRFWWDRyaWsL0znQU0a7Z7HTk2U7_AvNIVvH74vfswvkKr2RmE85Y-ikjMl637a5Cq45AoreUGcN3Jps-b3mBf5Sl9hZyuWsvE74WCBYCypgG'

// API ENDPOINTS:
const API_CREATE_PLAYLIST = `https://api.spotify.com/v1/users/${user_id}/playlists`;
const API_GET_CURRENT_TRACK = 'https://api.spotify.com/v1/me/player';


// Mehmet's custom function to handle json objects
function json_extractor(iterable_object, target_value){
    var result = [];
    for (i in iterable_object){
        result.push(iterable_object[i][target_value]);
    }
    return result.join(", ");

}

// SPOTIFY API FUNCTIONS
function create_playlist(name, public) {
    const ajaxRequest = new XMLHttpRequest();

    // Create a function that will receive data sent from the API
    ajaxRequest.onreadystatechange = function () {
        if (ajaxRequest.readyState == 4 && (ajaxRequest.status == 200 || ajaxRequest.status == 201)) {
            // request succesfull!
            //console.log(ajaxRequest.responseText);

            // convert the response into json
            var jsonObj = JSON.parse(ajaxRequest.responseText);
            console.log(jsonObj)

        } else if (ajaxRequest.readyState == 4) {
            console.log("Request Unsuccesful! \n", ajaxRequest)
        }
    }

    const json_data = {
        "name": name,
        "public": public
    };

    ajaxRequest.open("POST", API_CREATE_PLAYLIST, true);
    ajaxRequest.setRequestHeader('Authorization', `Bearer ${ACCESS_TOKEN}`);
    ajaxRequest.setRequestHeader('Content-Type', 'application/json');
    ajaxRequest.send(JSON.stringify(json_data));
    // JSON.stringify(json_data)
}

function get_current_track_info() {
    const ajaxRequest = new XMLHttpRequest();

    // Create a function that will receive data sent from the API
    ajaxRequest.onreadystatechange = function () {
        if (ajaxRequest.readyState == 4 && ajaxRequest.status == 200) {
            // request succesfull!
            //console.log(jsonObj)

            var jsonObj = JSON.parse(ajaxRequest.responseText);
            console.log(jsonObj);
            console.log(jsonObj.item.name);

            var song_id = jsonObj.item.id;
            var song_name = jsonObj.item.name;
            var song_link = jsonObj.item.external_urls.spotify;
            var artist_info = jsonObj.item.artists;
            var artist_names = json_extractor(artist_info, "name")

            document.getElementById("api_results").innerHTML = 'Song Name: ' + song_name + '<br><br> Artist Name:' + artist_names + '<br><br> Song Link: ' + song_link  ;


        } else if (ajaxRequest.readyState == 4 && ajaxRequest.status == 204) {
            console.log("Request Succesfull! \n But you're not listening to anything... \n", ajaxRequest)

        } else if (ajaxRequest.readyState == 4) {
            console.log("Request Unsuccesful! \n", ajaxRequest)
        }
    }

    ajaxRequest.open("GET", API_GET_CURRENT_TRACK, true);
    ajaxRequest.setRequestHeader('Authorization', `Bearer ${ACCESS_TOKEN}`);
    ajaxRequest.send();

}

function get_featured_playlists(){
    // TODO: ask Mehmet to update the remote repo!
}

function follow_artist(){
    // TODO: ask Mehmet to update the remote repo!
}

function follow_user(){
    // TODO: ask Mehmet to update the remote repo!
}

function follow_user(){
    // TODO: ask Mehmet to update the remote repo!
}


// ! DO NOT USE THIS FUNCTION !!!
function template_request(request_type, url, send_str = null) {
    // The variable that makes Ajax possible! 
    const ajaxRequest = new XMLHttpRequest();

    // Create a function that will receive data sent from the API
    ajaxRequest.onreadystatechange = function () {
        if (ajaxRequest.readyState == 4 && ajaxRequest.status == 200) {
            // request succesfull!
            //console.log(ajaxRequest.responseText);

            // convert the response into json
            var jsonObj = JSON.parse(ajaxRequest.responseText);
        } else if (ajaxRequest.readyState == 4) {
            console.log("Request Unsuccesful! \n", ajaxRequest)
        }
    }

    ajaxRequest.open(type, "AJAXdemo2.php" + queryString, true);
    ajaxRequest.setRequestHeader('Authorization', `Bearer ${ACCESS_TOKEN}`);
    ajaxRequest.send(send_str);

    // // Now get the value from user and pass it to server script.
    // var lastName = document.getElementById('lastName').value;
    // var firstName = document.getElementById('firstName').value;
    // var queryString = "?lastName=" + lastName ;
    // queryString +=  "&firstName=" + firstName + "&server=" + server + "&user=" + user + "&pwd=" + pwd + "&dbName=" + dbName;
    // ajaxRequest.open("GET", "AJAXdemo2.php" + queryString, true);
    // ajaxRequest.send(null);
}